src/cdsetool/download.py

```diff
@@ -26,6 +29,175 @@
 from cdsetool.monitor import NoopMonitor, StatusMonitor
 from cdsetool.query import FeatureQuery
 
+MANIFEST_BASENAMES = {
+    "SENTINEL-1": "manifest.safe",
+    "SENTINEL-2": "manifest.safe",
+    "SENTINEL-3": "manifest.xml",
+}
+
+
+def _href_to_url(odata_url: str, product_id: str, product_name: str, href: str) -> str:
+    """
+    Convert href, describing file location in manifest file, to an OData download URL.
+    """
+    path = "/".join([f"Nodes({item})" for item in href.split("/")])
+    return f"{odata_url}/Products({product_id})/Nodes({product_name})/{path}/$value"
+
+
+def filter_files(
+    manifest_file: str, pattern: str, exclude: bool = False
+) -> List[Dict[str, Any]]:
+    """
+    Filter a product's files, listed in its manifest, based on a given pattern.
+
+    Returns a list of files paths within the product bundle.
+
+    If "exclude" is set to False, only files that match pattern are returned. If it's
+    set to True, only files that do not match pattern are returned.
+    """
+    paths = []
+    xmldoc = etree.parse(manifest_file)
+
+    if os.path.basename(manifest_file) == "manifest.safe":
+        data_obj_section_elem = xmldoc.find("dataObjectSection")
+        for elem in data_obj_section_elem.iterfind("dataObject"):
+            path = elem.find("byteStream/fileLocation").attrib["href"]
+            path = path[2:]  # Remove "./" prefix present in S2 and S3 manifests
+            match = fnmatch.fnmatch(path.lower(), pattern)
+            if match and not exclude or exclude and not match:
```

Comment:
This looks like xor, `if match ^ exclude:`.

The `match = fnmatch..` and if-statement is identical between the two blocks, so share it:
```python
            path = "/".join(...)
    else:
        return paths
    match = fnmatch....
```